### PR TITLE
[WIP] High lvl commands

### PR DESCRIPTION
**Pull request by michalrudolf**

### CODE REVIEW

File: src/wui/wui.c [hunk 1]
```diff
@@ -54,24 +54,28 @@ static void update_wui_vars(void) {
...
 
-static int process_wui_request(char *request) {
-
-    if (strncmp(request, "!cip ", 5) == 0) {
-        uint32_t ip;
-        if (sscanf(request + 5, "%lu", &ip)) {
-            eeprom_set_var(EEVAR_CONNECT_IP4, variant8_ui32(ip));
+static int process_wui_request(wui_cmd_t *request) {
+
+    if (request->lvl == HIGH_LVL_CMD){
```
Comment (joshymjose) [inline src/wui/wui.c]: Does CONNECT need set command?

File: src/wui/wui_helper_funcs.c [hunk 2]
```diff
@@ -1,26 +1,25 @@
...
-#include "dbg.h"
-#include "wui.h"
 #include "string.h"
+#include "dbg.h"
 
 extern osMessageQId tcp_wui_mpool_id;
 extern osSemaphoreId tcp_wui_semaphore_id;
 
-void send_request_to_wui(const char *request) {
+uint8_t send_request_to_wui(wui_cmd_t * req_ptr) {
```
Comment (joshymjose) [inline src/wui/wui_helper_funcs.c]: why uint8_t ? why not uint32_t?

File: src/wui/wui_helper_funcs.c [hunk 3]
```diff
@@ -1,26 +1,25 @@
...
 
     osSemaphoreWait(tcp_wui_semaphore_id, osWaitForever);
     if (0 != tcp_wui_queue_id) // queue valid
-    {
+    {
```
Comment (joshymjose) [inline src/wui/wui_helper_funcs.c]: unwanted space here

File: src/wui/wui.c (hunk 1 above)
Comment (michalrudolf) [reply src/wui/wui.c]: I did not find any documentation on that topic. We only know, that CONNECT will need a SET feature to set all the customizable variables, detailes are matter of disscusion.. It's my design used on connect ip4, token and hostname.

File: src/wui/wui_helper_funcs.c (hunk 2 above)
Comment (michalrudolf) [reply src/wui/wui_helper_funcs.c]: ret val will be used to trigger "not enough spaci in osPool", so it should probably be HTTPC_COMMAND_STATUS instead

File: src/wui/wui.c [hunk 4]
```diff
@@ -54,24 +54,28 @@ static void update_wui_vars(void) {
...
+        if (request->high_lvl_cmd == CMD_SET){
+            if (request->var_msk & WUI_VAR_MSK(VAR_CONNECT_IP4)) {
+                uint32_t ip;
+                if (sscanf(request->arg, "%lu", &ip)) {
+                    eeprom_set_var(EEVAR_CONNECT_IP4, variant8_ui32(ip));
+                }
+            } else if (request->var_msk & WUI_VAR_MSK(VAR_CONNECT_TOKEN)) {
+                variant8_t token = variant8_pchar(request->arg, 0, 0);
```
Comment (joshymjose) [inline src/wui/wui.c]: i think this variant is not complete? anyway variant is something outside of WUI library

**Comment by joshymjose** [comment]
the parsing of low level command is missing!
implement ONLY ONE HIGH-LEVEL COMMAND NOW "send info"
now only send info in OTP memory firmware version, serial number etc"

File: src/wui/wui.c (hunk 4 above)
Comment (michalrudolf) [reply src/wui/wui.c]: You said to do it without the refactorings.. So it's without the refactorings

File: src/wui/wui_helper_funcs.c (hunk 2 above)
Comment (joshymjose) [reply src/wui/wui_helper_funcs.c]: i meant data type

File: src/common/wui_api.c [hunk 5]
```diff
@@ -6,6 +6,43 @@
  *      Author: joshy <joshymjose[at]gmail.com>
  */
 #include "wui_api.h"
+#include "version.h"
+#include "otp.h"
+#include <string.h>
+#include <stdio.h>
+
+#define PRINTER_TYPE_ADDR       0x0802002F  // 1 B
```
Comment (joshymjose) [inline src/common/wui_api.c]: should not redefine these address, might be already defined somewhere?

File: src/common/wui_api.c [hunk 6]
```diff
@@ -6,6 +6,43 @@
...
+void get_printer_info(printer_info_t * printer_info){
+    
+    // FIRMWARE VERSION
+    strlcpy(printer_info->firmware_version, project_version_full, FW_VER_STR_LEN);
+    // PRINTER TYPE
+    printer_info->printer_type = *(volatile uint8_t *)PRINTER_TYPE_ADDR;
+    // PRINTER_VERSION
+    printer_info->printer_version = *(volatile uint8_t *)PRINTER_VERSION_ADDR;
+    
+    char serial_numbers[OTP_SERIAL_NUMBER_SIZE];
```
Comment (joshymjose) [inline src/common/wui_api.c]: these local definitions are necessary?

File: src/wui/http/http_client.h [hunk 7]
```diff
@@ -44,7 +44,8 @@ typedef enum {
     CMD_REJT_CMD_ID,      // error with Command-Id
     CMD_REJT_CONT_TYPE,   // error with Content-Type
     CMD_REJT_GCODES_LIMI, // number of gcodes in x-gcode request exceeded
-    CMD_UNKNOWN,
+    CMD_INFO_REQ,         // special request - SEND_INFO
```
Comment (joshymjose) [inline src/wui/http/http_client.h:46]: this is not status of request, but a type of request

File: src/wui/wui_api.h [hunk 8]
```diff
@@ -13,6 +13,24 @@
...
 
+#include <stdint.h>
+
+#define FW_VER_STR_LEN      32
+#define MAC_ADDR_STR_LEN    18
+#define SER_NUM_STR_LEN     16
+#define UUID_STR_LEN        32
+#define PRI_STATE_STR_LEN   10
+typedef struct {
+
```
Comment (joshymjose) [inline src/wui/wui_api.h]: why you have here space

File: src/wui/wui_request_parser.c [hunk 9]
```diff
@@ -22,13 +23,7 @@ static int json_cmp(const char *json, jsmntok_t *tok, const char *s) {
...
 
 static HTTPC_COMMAND_STATUS parse_high_level_cmd(char *json, uint32_t len) {
-    char cmd_str[200];
-    uint32_t ret_code = httpc_json_parser(json, len, cmd_str);
-    if (ret_code) {
-        return CMD_REJT_CMD_STRUCT;
-    } else {
-        return CMD_ACCEPTED;
-    }
+    return httpc_json_parser(json, len);
```
Comment (joshymjose) [inline src/wui/wui_request_parser.c]: this redirection is unnecessary. just copy httpc_json_parsor code here

File: src/wui/wui_request_parser.c [hunk 10]
```diff
@@ -85,8 +100,7 @@ HTTPC_COMMAND_STATUS parse_http_reply(char *reply, uint32_t reply_len, httpc_hea
     return cmd_status;
 }
 
-uint32_t httpc_json_parser(char *json, uint32_t len, char *cmd_str) {
-    uint32_t ret_code = 1; // success is 0
+HTTPC_COMMAND_STATUS httpc_json_parser(char *json, uint32_t len) {
     int ret;
```
Comment (joshymjose) [inline src/wui/wui_request_parser.c]: if not necessary not use "int"  but uint32_t

File: src/wui/wui_request_parser.c [hunk 11]
```diff
@@ -97,49 +111,38 @@ uint32_t httpc_json_parser(char *json, uint32_t len, char *cmd_str) {
...
-            i++;
-        } else if (json_cmp(json, &t[i], "connect_name") == 0) {
-            strlcpy(request, json + t[i + 1].start, t[i + 1].end - t[i + 1].start + 1);
-            char connect_request[MAX_REQ_MARLIN_SIZE];
-            snprintf(connect_request, MAX_REQ_MARLIN_SIZE, "!cn %s", request);
-            send_request_to_wui(connect_request);
-            ret_code = 0;
-            i++;
+
+            ret_status = parse_high_cmd_args(&command, json, t, &i);
```
Comment (joshymjose) [inline src/wui/wui_request_parser.c]: i dont think this is a nice way to do this

Comment (joshymjose) [reply src/wui/wui_request_parser.c]: does SEND_INFO have some arguments? if not do not do anything now

**Comment by joshymjose** [comment]
1. All enums, Struct should have documentation of the variables
2. Functions should have documentation
   \brief breief info
then input parameters, direction of parameters: in, out or in/out
return value

File: src/common/wui_api.c (hunk 5 above)
Comment (michalrudolf) [reply src/common/wui_api.c]: I didn't find it

File: src/wui/wui_request_parser.c (hunk 10 above)
Comment (michalrudolf) [reply src/wui/wui_request_parser.c]: jsmn_parse() from jsmn.h returnes int.. That is why I use it there

**Pull request merged**